### PR TITLE
Movement

### PR DESCRIPTION
**Pull request by VitaJane**
Here's another change we made. It just adds a button to allow users to move the keyboard around.

### CODE REVIEW

**Comment by GreyWyvern** [comment]
This is great, and something I'd definitely like to add to the script. Thanks!

The main issue with it is that it doesn't account for the page scroll distance. If the viewport hasn't been scrolled at all, it works fine. But if the page has been scrolled down, then the positioning is off. (You can use `VKI_scrollDist()` to find this distance)

As well, as a "safety" measure, the movement event listener should also be removed from the document when the keyboard is closed for any reason. So something should be added in `this.VKI_close`.

*Edit: There should be a configuration option that can disable this also, but we can default it to "On".*

*Edit 2: The `mouseup` function should be attached to the document. If the user moves the mouse very fast before lifting the button, there's a good chance it will be outside the movement button which then won't receive the event.*

**Comment by VitaJane** [comment]
I'll get started adding those

**Comment by VitaJane** [comment]
Just pushed what I think will fix everything you suggested.

File: keyboard.js [hunk 1]
```diff
@@ -1704,6 +1705,35 @@ var VKI_attach, VKI_close;
           VKI_mouseEvents(numbkspspan);
           thth.appendChild(numbkspspan);
 
+        if (this.VKI_movement) {
+          function moveAt(pageX, pageY) {
+            this.VKI_keyboard.style.left = pageX + VKI_scrollDist()[0] - keyX + VKI_scrollDist()[0] + 'px';
+            this.VKI_keyboard.style.top = pageY + VKI_scrollDist()[1] - keyY + VKI_scrollDist()[1] + 'px';
+          }
```
Comment (GreyWyvern) [inline keyboard.js:1720]: These references should be `self.` to make explicit we are referencing the parent object.

*Edit: This should also take into account `self.VKI_target.keyboardPosition`.*

File: keyboard.js [hunk 2]
```diff
@@ -1704,6 +1705,35 @@ var VKI_attach, VKI_close;
...
+          let moveOsk = document.createElement('span');
+            moveOsk.id = 'moveOsk';
+            moveOsk.appendChild(document.createTextNode('\u2725'));
+            moveOsk.title = this.VKI_i18n['14'];
+            moveOsk.addEventListener('mousedown', function(event) {
+                event.preventDefault();
+                cord = document.getElementById('keyboardInputMaster').getBoundingClientRect();
+                keyX = event.pageX + VKI_scrollDist()[0] - cord.left;
+                keyY = event.pageY + VKI_scrollDist()[1] - cord.top;
```
Comment (GreyWyvern) [inline keyboard.js]: There needs to be a fork here based on `self.VKI_target.keyboardPosition == 'fixed'`. If this value is equal to `'fixed'` then scroll distance should be ignored.

File: keyboard.js [hunk 3]
```diff
@@ -1704,6 +1705,35 @@ var VKI_attach, VKI_close;
...
+                moveAt(event.pageX, event.pageY);
```
Comment (GreyWyvern) [inline keyboard.js]: Is this function call necessary? If I comment it out it still seems to work, although I might be missing an edge case?

**Review by GreyWyvern — CHANGES_REQUESTED** [review]
Looks and works nicely now! Just a few more small comments.

**Comment by VitaJane** [comment]
Ok, just pushed what should fix those issues, I think. You were right about the one function call being unnecessary. I think I fixed the issue with fixed too, but I'm not sure if I did that right, or understood it all the way.

**Review by GreyWyvern — APPROVED** [review]

**Comment by GreyWyvern** [comment]
This is looking good, and I can find no further issues with it. If everything's all ship-shape with you, I can merge this.

Thanks!

**Comment by VitaJane** [comment]
Sounds good to me!

**Pull request merged**